packages/eslint-plugin/src/rules/no-for-in-array.ts

```diff
@@ -45,3 +46,33 @@ export default createRule({
     };
   },
 });
+
+function isArgumentsObjectType(type: ts.Type): boolean {
+  return (
+    type.getSymbol()?.escapedName === ts.escapeLeadingUnderscores('IArguments')
+  );
+}
+
+function isRegExpExecArrayLike(program: ts.Program, type: ts.Type): boolean {
+  return isTypeRecurser(type, t =>
+    isBuiltinSymbolLike(program, t, 'RegExpExecArray'),
+  );
+}
+
+function isArray(checker: ts.TypeChecker, type: ts.Type): boolean {
+  return isTypeRecurser(
+    type,
+    t => checker.isArrayType(t) || checker.isTupleType(t),
+  );
+}
+
+function isTypeRecurser(
+  type: ts.Type,
+  predicate: (t: ts.Type) => boolean,
+): boolean {
+  if (type.isUnionOrIntersection()) {
+    return type.types.some(t => isTypeRecurser(t, predicate));
+  }
+
+  return predicate(type);
+}
```

Comment:
I've used this part from https://github.com/typescript-eslint/typescript-eslint/pull/10551.